x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java

```diff
@@ -0,0 +1,224 @@
+        ) Expression prefixLengthV6Field
+    ) {
+        super(
+            source,
+            prefixLengthV6Field == null
+                ? Arrays.asList(ipField, prefixLengthV4Field)
+                : Arrays.asList(ipField, prefixLengthV4Field, prefixLengthV6Field)
+        );
+        this.ipField = ipField;
+        this.prefixLengthV4Field = prefixLengthV4Field;
+        this.prefixLengthV6Field = prefixLengthV6Field;
+    }
+
+    public static IpPrefix readFrom(PlanStreamInput in) throws IOException {
+        return new IpPrefix(in.readSource(), in.readExpression(), in.readExpression(), in.readOptionalNamed(Expression.class));
+    }
+
+    public static void writeTo(PlanStreamOutput out, IpPrefix ipPrefix) throws IOException {
+        out.writeSource(ipPrefix.source());
+        List<Expression> fields = ipPrefix.children();
+        assert fields.size() == 2 || fields.size() == 3;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
+        out.writeOptionalWriteable(fields.size() == 3 ? o -> out.writeExpression(fields.get(2)) : null);
+    }
+
+    public Expression ipField() {
+        return ipField;
+    }
+
+    public Expression prefixLengthV4Field() {
+        return prefixLengthV4Field;
+    }
+
+    public Expression prefixLengthV6Field() {
+        return prefixLengthV6Field;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        var ipEvaluatorSupplier = toEvaluator.apply(ipField);
+        var prefixLengthV4EvaluatorSupplier = toEvaluator.apply(prefixLengthV4Field);
+
+        if (prefixLengthV6Field != null) {
+            var prefixLengthV6EvaluatorSupplier = toEvaluator.apply(prefixLengthV6Field);
+
+            return new IpPrefixEvaluator.Factory(
+                source(),
+                ipEvaluatorSupplier,
+                prefixLengthV4EvaluatorSupplier,
+                prefixLengthV6EvaluatorSupplier,
+                context -> new BytesRef(new byte[16])
+            );
+        }
+
+        return new IpPrefixOnlyV4Evaluator.Factory(
+            source(),
+            ipEvaluatorSupplier,
+            prefixLengthV4EvaluatorSupplier,
+            context -> new BytesRef(new byte[16])
+        );
+    }
+
+    @Evaluator(extraName = "OnlyV4")
+    static BytesRef process(BytesRef ip, int prefixLengthV4, @Fixed(includeInToString = false, build = true) BytesRef scratch) {
+        return process(ip, prefixLengthV4, 128, scratch);
+    }
+
+    @Evaluator
+    static BytesRef process(
+        BytesRef ip,
+        int prefixLengthV4,
+        int prefixLengthV6,
+        @Fixed(includeInToString = false, build = true) BytesRef scratch
+    ) {
+        if (ip.length != 16 || prefixLengthV4 < 0 || prefixLengthV4 > 32 || prefixLengthV6 < 0 || prefixLengthV6 > 128) {
```

Comment:
I think it's safe to assume `ip.length` is always 16. You could `assert` it if you like though. We just won't build this tree of evaluators if we aren't doing ips.

Interesting thing on the negative or long prefix lengths. `null` seems fine for now. Though we could throw an exception and warn on it.

Is it worth optimizing for the case when the lengths are constant? Not now, but maybe in a follow up. Not sure it'd be that much faster. A little, and it's nice practice for your first few contributions. But not worth blocking the PR at all.

Reply:
About the length check, the function cidr_match uses the length dynamically, as well as the CIDRUtils class (accepts 4 or 16 bytes). But maybe it's just for convenience. So removing the check for now.

Changed to an IllegalArgumentException.

About optimizing, you mean something like checking if it's a constant in the `toEvaluator()`, and using a custom evaluator//process or something like that? We can see later anyway yep

Reply:
> About optimizing, you mean something like checking if it's a constant in the toEvaluator(), and using a custom evaluator//process or something like that? We can see later anyway yep

Yeah, exactly that. I'm not sure if it'd be faster, but it could be. Might be worth trying later. Or not. No big deal.